java/vortex-jni/src/test/java/dev/vortex/api/TestMinimal.java

```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    @Test
+    public void testProjectedScan() {
+        var projectOptions = ScanOptions.builder().addColumns("Name", "State").build();
+
+        try (var file = Files.open(MINIMAL_PATH);
+                var projectedScan = file.newScan(projectOptions)) {
+            // Do stuff.
+            var dtype = projectedScan.getDataType();
+            assertEquals(DType.Variant.STRUCT, dtype.getVariant());
+            assertEquals(dtype.getFieldNames(), ImmutableList.of("Name", "State"));
+
+            var people = readToList(projectedScan, new TestCase() {
+                @Override
+                public Array[] open(Array batch) {
+                    return new Array[] {
+                        batch.getField(0), batch.getField(1),
+                    };
+                }
+
+                @Override
+                public Person readRow(Array[] fields, int idx) {
+                    return new Person(fields[0].getUTF8(idx), null, fields[1].getUTF8(idx));
+                }
+            });
+
+            assertEquals(PROJECTED_DATA, people);
+        }
+    }
+
+    @Test
+    public void testProjectedScanWithFilter() {
+        var filterOptions = ScanOptions.builder()
+                .addColumns("State", "Salary", "Name")
+                .predicate(Binary.eq(GetItem.of(Identity.INSTANCE, "State"), Literal.string("VA")))
+                .build();
+
+        try (var file = Files.open(MINIMAL_PATH);
+                var filteredScan = file.newScan(filterOptions)) {
+            var dtype = filteredScan.getDataType();
+            assertEquals(DType.Variant.STRUCT, dtype.getVariant());
+            assertEquals(dtype.getFieldNames(), ImmutableList.of("State", "Salary", "Name"));
+
+            var people = readToList(filteredScan, new TestCase() {
+                @Override
+                public Array[] open(Array batch) {
+                    return new Array[] {
+                        batch.getField(0), // state
+                        batch.getField(1), // salary
+                        batch.getField(2), // name
+                    };
+                }
+
+                @Override
+                public Person readRow(Array[] fields, int idx) {
+                    var state = fields[0].getUTF8(idx);
+                    var salary = fields[1].getBigDecimal(idx);
+                    var name = fields[2].getUTF8(idx);
+                    return new Person(name, salary, state);
+                }
+            });
+
+            assertEquals(ImmutableList.of(new Person("John", BigDecimal.valueOf(10_000L, 2), "VA")), people);
+        }
+    }
+
+    interface TestCase {
+        Array[] open(Array batch);
+
+        Person readRow(Array[] fields, int idx);
+    }
+
+    private List<Person> readToList(ArrayIterator scan, TestCase testCase) {
+        ImmutableList.Builder<Person> people = ImmutableList.builder();
+        while (scan.hasNext()) {
+            var batch = scan.next();
```

Comment:
this is one case where spliterator is useful. Anyway, we can file this as an improvement